### DISCUSSION TITLE
Avoid opening a new thread for TruffleCompiler

### DISCUSSION THREAD

**Discussion by nsoubelet** [oracle/graal#9045] [Q&A] [answered]
Hello team,
I recently noticed that when evaluating a JavaScript source, Truffle enqueues the operation and opens a new thread called `TruffleCompilerThread`, something I presume happens [here](https://github.com/oracle/graal/blob/1d06946eb89f388855988da610a27b27dfca08db/truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/BackgroundCompileQueue.java).

Since on my app's side the evaluation is done already in a separate Thread, I would like to know whether it is possible to tell Truffle to avoid opening a new one but use the current instead.
First look suggests it is not possible since it must be always enqueued, but probably I'm missing something.

Thanks!

**Comment by chumer** [comment, marked as answer]
Hi!

Yes, you can perform the compilation on the main thread by setting the `engine.BackgroundCompilation` option to `false`. However, we do not recommend doing this, as it can cause latency spikes in the evaluation of code (up to seconds). I have yet to encounter a use case where I would recommend turning off background compilation.

The compilation is expected to be performed only once per polyglot `Source`, so if you are running in a long-running server process, I would recommend keeping background compilation on, but at the same time making sure code caching across contexts is enabled, so scripts only need to be compiled once per server process. See here for instructions: 
https://www.graalvm.org/latest/reference-manual/embed-languages/#code-caching-across-multiple-contexts

If your scripts are short-lived and very simple, it can make sense to turn off compilation entirely by setting the `engine.Compilation` option to `false`. With this option set to `false` Truffle will never create a compiler thread.

If the compiler uses too much CPU resources, you can, for example, reduce the number of compiler threads by setting `engine.CompilerThreads` to `1`. By default, it selects a number depending on the number of available CPU cores.

More info on setting options: https://www.graalvm.org/latest/reference-manual/polyglot-programming/#polyglot-options
A full list of available options can be found here: https://github.com/oracle/graal/blob/master/truffle/docs/Options.md

Hope this helps.

**Reply by nsoubelet** [reply]
Hello @chumer,
Thank you for the detailed answer. That helps for sure.

Unfortunately, my scripts are short-lived but can be complex (using _resources-needed+lines-of-code_ as a way of measuring complexity), a rare combination. So, yes, disabling `engine.BackgroundCompilation` can be potentially problematic.
I will opt then for caching `Source`s as the number is deterministic and, if I find something interesting, will share some benchmark with you.

**Comment by nsoubelet** [comment]
Hello @chumer, bothering you with this subject again.

I finally implemented a caching mechanism for `Source` but I still see several `TruffleCompilerThread` being created.
After some investigation, and due to the fact that I implemented a custom `FileSystem`, I noticed that when the `Source` is evaluated, all the `import`s are compiled "again".

Hence my question: is it possible to create a `Source` with the full tree, that is, including its dependencies, so we avoid getting them from the `FileSystem` and, more important, recompiling them each time?
Note: `Source.Builder` has `cache` flag enabled.

Thanks!
